File: common/src/main/java/dev/aika/bigclever/tunnel/TunnelManager.java
package dev.aika.bigclever.tunnel;


import dev.aika.bigclever.config.BigCleverConfig;
import dev.aika.bigclever.utils.DNSQuery;

import java.util.ArrayList;
import java.util.List;

public class TunnelManager {
    private final List<Tunnel> activeTunnels = new ArrayList<>();
    private BigCleverConfig config;

    private boolean wstunnelWaitForDownload = false;

    public void checkTunnels() {
        config = BigCleverConfig.get();

        wstunnelWaitForDownload = config.wstunnelEnable && !new WSTunnel(null).isAvailable();
    }

    public Tunnel useTunnel(String host, int port, boolean mcPortOnly) {
        if (!DNSQuery.isSuitableForDNSQuery(host)) return null;
        for (var record : DNSQuery.queryTXTRecords(host)) {
            if (wstunnelWaitForDownload || !config.wstunnelEnable) continue;

            Tunnel tunnel = WSTunnel.fromRecord(record);
            if (tunnel.isAvailable()) {
                tunnel.start(host, port, mcPortOnly);
                if (tunnel.getServerAddress() == null) {
                    tunnel.stop();
                    continue;
                }
                activeTunnels.add(tunnel);
                return tunnel;
            }
        }
        return null;
    }

    public void closeAll() {
        for (var tunnel : activeTunnels) tunnel.stop();
    }
}

File: common/src/main/kotlin/dev/aika/bigclever/tunnel/TunnelManager.kt
package dev.aika.bigclever.tunnel

import dev.aika.bigclever.BigClever
import dev.aika.bigclever.config.BigCleverConfig.Companion.config
import dev.aika.bigclever.utils.DNSQuery

class TunnelManager {
    private val activeTunnels: MutableList<Tunnel> = ArrayList()

    private var wstunnelWaitForDownload = false

    fun checkTunnels(): Boolean {
        wstunnelWaitForDownload = config.wstunnelEnable && !WSTunnel(arrayOf()).isAvailable
        if (wstunnelWaitForDownload) BigClever.LOGGER.warn("wstunnel is not available, waiting for download...")

        return wstunnelWaitForDownload
    }

    fun useTunnel(host: String, port: Int, mcPortOnly: Boolean): Tunnel? {
        if (!DNSQuery.isSuitableForDNSQuery(host)) return null

        for (record in DNSQuery.queryTXTRecords(host)) {
            if (wstunnelWaitForDownload || !config.wstunnelEnable) continue

            val tunnel: WSTunnel = WSTunnel.fromRecord(record)
            if (tunnel.isAvailable) {
                tunnel.startClient(host, port, mcPortOnly)
                if (tunnel.serverAddress == null) {
                    tunnel.stop()
                    continue
                }
                activeTunnels.add(tunnel)
                return tunnel
            }
        }
        return null
    }

    fun closeAll() {
        for (tunnel in activeTunnels) tunnel.stop()
    }
}